Авто тесты Веб-приложения v1.0

CRASH3000 opened the pull request:
### Описание

Добавлен проект для автоматизации тестирования веб-приложения **Stellar Burgers**. Включает тесты для проверки основных функций приложения, а также кроссбраузерное тестирование в браузерах Google Chrome и Firefox. В проекте реализованы паттерны Page Object Model и подключен Allure для генерации отчетов.

### Основные изменения
- Добавлен файл README.md с инструкциями по установке, запуску тестов и генерации отчетов.
- Настроены тестовые данные и URL-адреса в директории **data/**.
- Реализована структура Page Object в директории **page_objects/**.
- Добавлены тестовые сценарии для разделов восстановления пароля, личного кабинета, основного функционала и ленты заказов.
- Настроены зависимости в **requirements.txt**.

### Как запустить
1. Установите зависимости из **requirements.txt**.
2. Запустите тесты командой `pytest --alluredir=allure-results`.
3. Сгенерируйте Allure отчет командой `allure serve allure-results`.

### Дополнительно
- В проекте реализовано кроссбраузерное тестирование с поддержкой Firefox (необходим Gekodriver).

### Скриншоты Allure-отчёта 
![Снимок экрана 2024-11-07 в 22 50 15](https://github.com/user-attachments/assets/aee5eafe-065f-45c7-a382-7f0ca5646f6d)
![Снимок экрана 2024-11-07 в 22 50 57](https://github.com/user-attachments/assets/18ae8129-619d-45d3-954b-f586ab362034)

CRASH3000 committed:
Авто тесты Веб-приложения v1.0
Настроена архитектура проекта, структура тестов и README для автоматизации Stellar Burgers

Что было сделано?
- Создан проект для автоматизации тестирования веб-приложения Stellar Burgers.
- Написаны автотесты для проверки основных функций приложения.
- Настроены кроссбраузерные тесты для Google Chrome и Firefox.
- Реализован паттерн Page Object Model.
- Подключен Allure для генерации отчетов.

AlexKitIt commented on data/urls_site_data.py:
```diff
@@ -0,0 +1,10 @@
+
+    ORDER_FEED_URL = "https://stellarburgers.nomoreparties.site/feed"
+
+    LOGIN_URL = "https://stellarburgers.nomoreparties.site/login"
+    FORGOT_PASSWORD_URL = "https://stellarburgers.nomoreparties.site/forgot-password"
+    RESET_PASSWORD_URL = "https://stellarburgers.nomoreparties.site/reset-password"
+    PROFILE_URL = "https://stellarburgers.nomoreparties.site/account/profile"
+    ORDER_HISTORY_URL = "https://stellarburgers.nomoreparties.site/account/order-history"
```
Можно лучше: урлы стоит конкатенировать с базовым, так будет проще поддерживать несколько стендов. Записать в переменную baseUrl и прикручивать к ней эндпоинты

CRASH3000 replied:
✅ Правки внесены.

AlexKitIt commented on page_objects/login_page.py:
```diff
@@ -0,0 +1,50 @@
+from selenium.webdriver.support import expected_conditions as EC
+from data.urls_site_data import UrlsSiteData
+
+class LoginPage:
```
В работе остутствует BasePage, в котором будут базовые методы для взаимодействия со всеми элементами по типу клика или send_keys и базовые wait, которых уже не должно быть в объектах страниц, 
там, за счет наследования объектов от Base, можно будет эти общие методы вызывать уже для элементов страниц. Вот эта модель и будет PageObject. 
Все обращения к driver - в BasePage, все работа с конкретными элементами на страницах - в объектах страниц

CRASH3000 replied:
✅ Правки внесены.

AlexKitIt commented on page_objects/login_page.py:
```diff
@@ -0,0 +1,50 @@
+        self.driver = driver
+        self.wait = WebDriverWait(driver, 10)
+
+    def open_login_page(self):
```
Нужно исправить: здесь и далее: над каждым методом пейджи ставится аннотация allure.step

CRASH3000 replied:
✅ Правки внесены.

AlexKitIt commented on page_objects/login_page.py:
```diff
@@ -0,0 +1,50 @@
+        self.wait = WebDriverWait(driver, 10)
+
+    def open_login_page(self):
+        self.driver.get(UrlsSiteData.LOGIN_URL)
+
+    def open_restore_password_page(self):
+        self.driver.get(UrlsSiteData.RESET_PASSWORD_URL)
+
+    def wait_restore_password_page(self):
+        self.wait.until(
+            EC.url_to_be(UrlsSiteData.RESET_PASSWORD_URL)
+        )
```
Нужно исправить: здесь и далее по проекту: все прямые обращения к driver и waitы необходимо вынести в base, там описать методы взаимодействия с ними и оттуда за счет наследования классов страниц от base переиспользовать

CRASH3000 replied:
✅ Правки внесены.

AlexKitIt commented on page_objects/main_page.py:
```diff
@@ -0,0 +1,142 @@
+from data.urls_site_data import UrlsSiteData
+from selenium.common.exceptions import ElementClickInterceptedException
+
+class MainPage:
```
Нужно исправить: здесь и далее: все пейджи наследуются от base

CRASH3000 replied:
✅ Правки внесены.

AlexKitIt commented on tests/conftest.py:
```diff
@@ -0,0 +1,44 @@
+            firefox_options = webdriver.FirefoxOptions()
+            # Указываем прямой путь к geckodriver
+            geckodriver_path = "/usr/local/bin/geckodriver"
+            driver = webdriver.Firefox(service=FirefoxService(executable_path=geckodriver_path),
```
executable_path использовать нельзя, в CI не заведется

CRASH3000 replied:
✅ Правки внесены.

AlexKitIt commented on tests/test_02_profile.py:
```diff
@@ -0,0 +1,56 @@
+        main_page.open_main_page()
+        main_page.click_account_button()
+
+        assert driver.current_url == UrlsSiteData.LOGIN_URL, "Не удалось перейти на страницу авторизации"
```
Нужно исправить: здесь и далее: прямого обращения к driver, вызова его property, execute_script не нем и тд из тестов быть не должно. Для этого в base пишутся соответствующие методы и тут вызываются на объектах конкретных страниц

CRASH3000 replied:
✅ Правки внесены.

CRASH3000 committed:
Авто тесты Веб-приложения v1.1
Обновление проекта с исправлениями и улучшениями на основе ревью

Что было сделано?
- Унификация ссылок для упрощения поддержки на нескольких стендов.
- Создан базовый класс BasePage. Сюда вынесены общие методы для взаимодействия с элементами во время тестирования. Все страницы теперь наследуются от BasePage
- Добавлены аннотации allure.step для каждого метода страниц.
- Убраны все прямые обращения к драйверу (driver.get, driver.find_element) из тестов.
- Настроен webdriver-manager для Firefox (С сохранением второго способа использования Gecko Driver)
- Отказ от прямого обращения к driver в тестах.

AlexKitIt commented on page_objects/login_page.py:
```diff
@@ -0,0 +1,39 @@
+    def __init__(self, driver):
+        super().__init__(driver)
+
+    def open_login_page(self):
```
Нужно исправить: здесь и далее:  декоратор step не добавлен

CRASH3000 replied:
✅ Правки внесены.

AlexKitIt commented on page_objects/login_page.py:
```diff
@@ -0,0 +1,39 @@
+from locators.login_page_locators import LoginPageLocators
+from selenium.webdriver.support.ui import WebDriverWait
+from selenium.webdriver.support import expected_conditions as EC
```
Нужно исправить: здесь и далее: неиспользуемый импорты

CRASH3000 replied:
✅ Правки внесены.

AlexKitIt commented on page_objects/main_page.py:
```diff
@@ -0,0 +1,126 @@
+
+    @allure.step("Перетаскиваем ингредиент в зону 'Выбранные ингредиенты'")
+    def add_ingredient_to_order(self):
+        ingredient = self.driver.find_element(*MainPageLocators.INGREDIENT_ITEM)
```
Повторный коммент: Нужно исправить: здесь и далее по проекту: все прямые обращения к driver и waitы EC и тд необходимо вынести в base, там описать методы взаимодействия с ними и оттуда за счет наследования классов страниц от base переиспользовать

CRASH3000 replied:
✅ Правки внесены.

AlexKitIt commented on tests/test_03_main_features.py:
```diff
@@ -0,0 +1,93 @@
+        main_page.click_constructor()
+
+        # Проверка URL
+        assert driver.current_url == UrlsSiteData.MAIN_URL
```
Повторный коммент: Нужно исправить: здесь и далее: прямого обращения к driver, вызова его property, execute_script не нем и тд из тестов быть не должно. Для этого в base пишутся соответствующие методы и тут вызываются на объектах конкретных страниц

CRASH3000 replied:
✅ Правки внесены.

CRASH3000 committed:
Авто тесты Веб-приложения v1.2
Обновление проекта с исправлениями и улучшениями на основе ревью

Что было сделано?
- Добавил декодеры step в page_objects/login_page.py;
- Удалил неиспользуемый импорты;
- Все прямые обращения к driver и wait EC и тд вынес в base;
- Убрал прямое обращения к driver, вызова его property, execute_script и тд из тестов;
- Проверил работу тестов после правок кода.
